tests/spin/test_spin_hamiltonian.py

```diff
@@ -782,3 +783,826 @@ def test_fermi_hubbard_hamiltonian_matrix(shape, n_cells, t, coulomb, expected_h
+            + (1.1556861568115007 + 0j) * (X(2) @ Z(3) @ Y(4))
+            + (-1.1556861568115007 + 0j) * (Y(3) @ Z(4) @ X(5))
+            + (-0.182454122875488 + 0j) * (Y(3) @ Z(4) @ Y(5))
+            + (-0.182454122875488 + 0j) * (X(3) @ Z(4) @ X(5))
+            + (1.1556861568115007 + 0j) * (X(3) @ Z(4) @ Y(5)),
+        ),
+        (
+            "square",
+            [2, 2, 1],
+            [[0, 0.5, 0.5, 0], [0.5, 0, 0, 0.5], [0.5, 0, 0, 0.5], [0, 0.5, 0.5, 0]],
+            1.0,
+            [
+                [0.0, 1.0, 1.0, 1.41421356],
+                [1.0, 0.0, 1.41421356, 1.0],
+                [1.0, 1.41421356, 0.0, 1.0],
+                [1.41421356, 1.0, 1.0, 0.0],
+            ],
+            False,
+            (-0.25 + 0j) * (Y(0) @ Z(1) @ Y(2))
+            + (-0.25 + 0j) * (X(0) @ Z(1) @ X(2))
+            + (-0.25 + 0j) * (Y(1) @ Z(2) @ Y(3))
+            + (-0.25 + 0j) * (X(1) @ Z(2) @ X(3))
+            + (-0.25 + 0j) * (Y(0) @ Z(1) @ Z(2) @ Z(3) @ Y(4))
+            + (-0.25 + 0j) * (X(0) @ Z(1) @ Z(2) @ Z(3) @ X(4))
+            + (-0.25 + 0j) * (Y(1) @ Z(2) @ Z(3) @ Z(4) @ Y(5))
+            + (-0.25 + 0j) * (X(1) @ Z(2) @ Z(3) @ Z(4) @ X(5))
+            + (-0.25 + 0j) * (Y(2) @ Z(3) @ Z(4) @ Z(5) @ Y(6))
+            + (-0.25 + 0j) * (X(2) @ Z(3) @ Z(4) @ Z(5) @ X(6))
+            + (-0.25 + 0j) * (Y(3) @ Z(4) @ Z(5) @ Z(6) @ Y(7))
+            + (-0.25 + 0j) * (X(3) @ Z(4) @ Z(5) @ Z(6) @ X(7))
+            + (-0.25 + 0j) * (Y(4) @ Z(5) @ Y(6))
+            + (-0.25 + 0j) * (X(4) @ Z(5) @ X(6))
+            + (-0.25 + 0j) * (Y(5) @ Z(6) @ Y(7))
+            + (-0.25 + 0j) * (X(5) @ Z(6) @ X(7))
+            + (-0.4938829729963678 + 0j) * (Y(0) @ Z(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ X(6))
+            + (-0.07797184738268718 + 0j) * (Y(0) @ Z(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ Y(6))
+            + (-0.07797184738268718 + 0j) * (X(0) @ Z(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ X(6))
+            + (0.4938829729963678 + 0j) * (X(0) @ Z(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ Y(6))
+            + (-0.4938829729963678 + 0j) * (Y(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ Z(6) @ X(7))
+            + (-0.07797184738268718 + 0j) * (Y(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ Z(6) @ Y(7))
+            + (-0.07797184738268718 + 0j) * (X(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ Z(6) @ X(7))
+            + (0.4938829729963678 + 0j) * (X(1) @ Z(2) @ Z(3) @ Z(4) @ Z(5) @ Z(6) @ Y(7))
+            + (-0.4938829729963678 + 0j) * (Y(2) @ Z(3) @ X(4))
+            + (-0.07797184738268718 + 0j) * (Y(2) @ Z(3) @ Y(4))
+            + (-0.07797184738268718 + 0j) * (X(2) @ Z(3) @ X(4))
+            + (0.4938829729963678 + 0j) * (X(2) @ Z(3) @ Y(4))
+            + (-0.4938829729963678 + 0j) * (Y(3) @ Z(4) @ X(5))
+            + (-0.07797184738268718 + 0j) * (Y(3) @ Z(4) @ Y(5))
+            + (-0.07797184738268718 + 0j) * (X(3) @ Z(4) @ X(5))
+            + (0.4938829729963678 + 0j) * (X(3) @ Z(4) @ Y(5)),
+        ),
+        (
+            "square",
+            [2, 2, 1],
```

Comment:
This test also seem a bit repetitive.

Reply:
This is the only test which tests the Hamiltonian when all of hopping1, hopping2, and phi are provided as a matrix/list instead of being constant.